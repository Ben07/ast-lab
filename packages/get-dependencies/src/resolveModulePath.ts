import enhancedResolve from 'enhanced-resolve';
import path from 'path';

import isCore from './isCore';
import isThirdParty from './isThirdParty';
import { Alias } from './types';

type ResolveOptions = {
  alias?: Alias,
  moduleDirectory?: string[],
  extensions?: string[],
  plugins?: any[]
}

export default async function resolveModulePath(
  mod: string, 
  source: string, 
  { 
    alias, 
    moduleDirectory: modules, 
    extensions,
    plugins
  }: ResolveOptions = {}
): Promise<string | void> {
  if (!mod || isCore(mod)) {
    return;
  }
  if (path.isAbsolute(mod)) {
    return mod;
  }
  const resolver = enhancedResolve.create({
    extensions: extensions,
    modules,
    alias,
    plugins
  });
  const basedir = path.dirname(source);
  const result = await new Promise((resolve, reject) => resolver(basedir, mod, (err: Error, res: string | undefined) => {
    if (err) {
      return reject(err.message);
    }
    if (!res) {
      console.log(`Couldn't find module ${mod}!`);
      return resolve();
    }
    if (isThirdParty(res)) {
      return resolve();
    }
    return resolve(res);
  })) as string | void;
  return result;
}
