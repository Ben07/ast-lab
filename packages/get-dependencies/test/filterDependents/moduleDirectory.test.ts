// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import { filterDependents } from './utils';
import resolveModulePath from '../../src/createResolver';

enumerateArgsTestFunction(
  filterDependents.bind(null, 'module-directory'),
  configArgs()
  .arg('sources', [
    ['a.js', 'b.js'],
  ])
  .arg('targets', [
    { './a': ['default'] }
  ])
  .arg('options', [
    {
      resolver: resolveModulePath({
        modules: ['../']
      })
    }
  ]),
  'module alias'
);