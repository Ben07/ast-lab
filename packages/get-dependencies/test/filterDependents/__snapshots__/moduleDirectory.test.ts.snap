// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module alias - (["sources", "targets", "options"]) - [["a.js","b.js"],{"./a":["default"]},{}] 1`] = `
Array [
  "/module-directory/a.js",
  "/module-directory/b.js",
]
`;
